### ISSUE TITLE
Draw Interaction result not showing with stopClick and quick double click to close

### ISSUE THREAD

**Issue by EvertEt** [openlayers/openlayers#12171]
**Describe the bug**
When using a Draw interaction with stopClick, closing the polygon with a quick double click, the drawing disappears

**To Reproduce**
1. Go to https://codesandbox.io/s/draw-features-forked-ob1nz?file=/main.js
2. Draw a polygon and close it quickly with a double click
3. See polygon not showing until map move

**Expected behavior**
The drawn polygon is shown directly

https://user-images.githubusercontent.com/9197891/113266147-32a07d00-92d5-11eb-9d73-b7dddb795b53.mov

Might very well be related to #12161

**Comment by ahocevar** [comment]
I also think this might be related to #12161. And maybe it makes the problem easier to debug. But I cannot reproduce it either.

**Comment by EvertEt** [comment]
I am using a touchpad and can't reproduce it as well so I suppose having a mouse (which I guess is faster) is a big factor.

**Comment by EvertEt** [comment]
On both issues, we are open to do some suggested tests with code changes to provide better debugging info.

**Comment by EvertEt** [comment]
What's even weirder, it that we noticed this issue but we use the draw interaction differently: on drawend we add the feature (through react) to a vectorLayer and there it's also showing the same behaviour

**Comment by ahocevar** [comment]
Can you try this change?
```diff
--- a/src/ol/MapBrowserEventHandler.js
+++ b/src/ol/MapBrowserEventHandler.js
@@ -122,17 +122,20 @@ class MapBrowserEventHandler extends Target {
   }
 
   /**
-   * @param {PointerEvent} pointerEvent Pointer
+   * @param {PointerEvent} pointerEvent Pointer event
+   * @param {MapBrowserEvent} upEvent Pointerup event
    * event.
    * @private
    */
-  emulateClick_(pointerEvent) {
+  emulateClick_(pointerEvent, upEvent) {
     let newEvent = new MapBrowserEvent(
       MapBrowserEventType.CLICK,
       this.map_,
       pointerEvent
     );
-    this.dispatchEvent(newEvent);
+    if (!upEvent.propagationStopped) {
+      this.dispatchEvent(newEvent);
+    }
     if (this.clickTimeoutId_ !== undefined) {
       // double-click
       clearTimeout(this.clickTimeoutId_);
@@ -142,7 +145,9 @@ class MapBrowserEventHandler extends Target {
         this.map_,
         pointerEvent
       );
-      this.dispatchEvent(newEvent);
+      if (!upEvent.propagationStopped) {
+        this.dispatchEvent(newEvent);
+      }
     } else {
       // click
       this.clickTimeoutId_ = setTimeout(
@@ -154,7 +159,9 @@ class MapBrowserEventHandler extends Target {
             this.map_,
             pointerEvent
           );
-          this.dispatchEvent(newEvent);
+          if (!upEvent.propagationStopped) {
+            this.dispatchEvent(newEvent);
+          }
         }.bind(this),
         250
       );
@@ -204,11 +211,10 @@ class MapBrowserEventHandler extends Target {
     // event.stopPropagation() or event.preventDefault().
     if (
       this.emulateClicks_ &&
-      !newEvent.propagationStopped &&
       !this.dragging_ &&
       this.isMouseActionButton_(pointerEvent)
     ) {
-      this.emulateClick_(this.down_);
+      this.emulateClick_(this.down_, newEvent);
     }
 
     if (this.activePointers_ === 0) {
```

**Comment by BerndSchrooten** [comment]
@ahocevar I tried out your change and it has the same results. In both cases (label rendered immediately and label rendered only after moving the map) the POINTERUP event has `stopPropagation` set to `true`

**Comment by ahocevar** [comment]
Ok, now can you please try this change:
```js
--- a/src/ol/interaction/Draw.js
+++ b/src/ol/interaction/Draw.js
@@ -650,7 +650,7 @@ class Draw extends PointerInteraction {
     }
 
     if (!pass && this.stopClick_) {
-      event.stopPropagation();
+      event.propagationStopped = true;
     }
     return pass;
   }
```

**Comment by BerndSchrooten** [comment]
Same behavior, I applied it on top of your first proposed changes as well as on its own.

**Comment by ahocevar** [comment]
Ok, then I'm out of ideas again - at least if the problem really only occurs when `stopClick` is true.

**Comment by EvertEt** [comment]
Another curious behaviour: Whenever a drawing is hidden, all next drawings will be hidden as well (as slow as they might be) until the map is moved and they all appear.

**Comment by EvertEt** [comment]
Is there some behaviour for not updating layers while dragging or something alike?

**Comment by MoonE** [comment]
@EvertEt https://openlayers.org/en/main/apidoc/module-ol_layer_VectorTile-VectorTileLayer.html `updateWhileInteracting` / `updateWhileAnimating`

**Comment by EvertEt** [comment]
Thanks @MoonE ! Adding logs (in VectorLayer prepareFrame) shows that the map is stuck in `interacting` (`frameState.viewHints[ViewHint.INTERACTING] === 1`). The weird thing is, the Draw interaction VectorLayer has updateWhileInteracting set to true.

But this indeed seems the cause, executing `temp1.getView().setHint(1,-1)` (Unsetting the frameState isinteracting) does render the hidden features!

**Comment by EvertEt** [comment]
So the DragPan receives a `pointerdrag` event, but no `pointerup` so the map stays in "interacting" mode (ref #12177). This also is not necessarily on the double click to finish.

This would also explain why it only happens with `stopClick` since the `pointup` event will be blocked in this case.

**Comment by kikitte** [comment]
The below picture is the logs about what interactions handle mapBrowserEvent with `stopClick`  set to true on the drawer, 
![image-20210404113938217](https://user-images.githubusercontent.com/23274684/113506858-c6927300-9579-11eb-8dfe-49ea0aa0bccf.png)
Only the drawer interaction handles the pointer-up event because the propagation is set to false in this code https://github.com/openlayers/openlayers/blob/6f7b6c8141ed0792725002b88519daaec6f3b9fb/src/ol/interaction/Draw.js#L652-L654.
So the DragPan set the view to be interacting but without recovering from it.

And I have tried to set the propagation to true in handleDownEvent in Drawer.js just like handleUpEvent does, so the DragPan will not set the view to be interacting. This problem looks fixed but we can't drag the view anymore during drawing after doing this.

**Comment by ahocevar** [comment]
Note that line 653 above stops propagation on the OpenLayers event and the native pointer event. This is why I had suggested to try changing that line to `event.propagationStopped = true`, because in that case the native event chain would not be affected. I'm actually surprised that this did not fix the issue.

**Comment by EvertEt** [comment]
I also verified it does not fix the issue. The `handleUpEvent` of the DragPan seems to be triggered on an OL event, not a native one. Where do you think the native event would still have an impact?

Is the bug here that the DragPan is beginning an interaction during drawing, or is the bug that it does not finish?

**Comment by ahocevar** [comment]
Thanks for sharing so many insights here. I'm pretty confident now that #12185 fixes this issue. @EvertEt, please test.

**Comment by EvertEt** [comment]
It does indeed seem to fix it on my end, amazing work @ahocevar ! Do you still believe it is related to #12161 and this one should also be fixed by #12185?

@kikitte @BerndSchrooten Can you also verify #12185 (`ahocevar:prevent-default`) fixes the problems?

**Comment by kikitte** [comment]
@EvertEt Yes, it works for me too.

**Comment by ahocevar** [comment]
@EvertEt 

> Do you still believe it is related to #12161 and this one should also be fixed by #12185?

I'm not 100% sure, but can you please test?

**Comment by BerndSchrooten** [comment]
It looks like that fixes this issue! 👍
